core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java

```diff
@@ -364,9 +442,39 @@ private Certificate fetchEphemeralCertificate(KeyPair keyPair) {
               connectionName),
           ex);
     }
+    // Update certExpiration with the value for the new cert
+    X509Certificate x509Certificate = (X509Certificate) ephemeralCertificate;
+    certExpiration = x509Certificate.getNotAfter();
+
     return ephemeralCertificate;
   }
 
+  private Date getTokenExpirationTime() {
+    return tokenSource.getAccessToken().getExpirationTime();
+  }
+
+  private long secondsUntilRefresh() {
+    Duration refreshBuffer = enableIamAuth ? iamAuthRefreshBuffer : defaultRefreshBuffer;
+    if (enableIamAuth) {
+      Date tokenExpiration = getTokenExpirationTime();
+      if (certExpiration.after(tokenExpiration)) {
+        certExpiration = tokenExpiration;
+      }
+    }
+
+    Duration timeUntilRefresh = Duration.between(Instant.now(), certExpiration.toInstant())
+        .minus(refreshBuffer);
+
+    if (timeUntilRefresh.isNegative()) {
+      // If the time until the certificate expires is less than the buffer, schedule the refresh
+      // closer to the expiration time
+      timeUntilRefresh = Duration.between(Instant.now(), certExpiration.toInstant())
+          .minus(Duration.ofSeconds(5));
+    }
+    return timeUntilRefresh.getSeconds();
+  }
```

Comment:
It seems like it would be better to store this data in the InstanceData object, so that it's calculated once